### https://leetcode.com/problems/minimum-speed-to-arrive-on-time/
# [1870. 准时到达的列车最小时速](https://leetcode.cn/problems/minimum-speed-to-arrive-on-time)

[English Version](/solution/1800-1899/1870.Minimum%20Speed%20to%20Arrive%20on%20Time/README_EN.md)

## 题目描述

<!-- 这里写题目描述 -->

<p>给你一个浮点数 <code>hour</code> ，表示你到达办公室可用的总通勤时间。要到达办公室，你必须按给定次序乘坐 <code>n</code> 趟列车。另给你一个长度为 <code>n</code> 的整数数组 <code>dist</code> ，其中 <code>dist[i]</code> 表示第 <code>i</code> 趟列车的行驶距离（单位是千米）。</p>

<p>每趟列车均只能在整点发车，所以你可能需要在两趟列车之间等待一段时间。</p>

<ul>
	<li>例如，第 <code>1</code> 趟列车需要 <code>1.5</code> 小时，那你必须再等待 <code>0.5</code> 小时，搭乘在第 2 小时发车的第 <code>2</code> 趟列车。</li>
</ul>

<p>返回能满足你准时到达办公室所要求全部列车的<strong> 最小正整数 </strong>时速（单位：千米每小时），如果无法准时到达，则返回 <code>-1</code> 。</p>

<p>生成的测试用例保证答案不超过 <code>10<sup>7</sup></code> ，且 <code>hour</code> 的 <strong>小数点后最多存在两位数字</strong> 。</p>

<p> </p>

<p><strong>示例 1：</strong></p>

<pre>
<strong>输入：</strong>dist = [1,3,2], hour = 6
<strong>输出：</strong>1
<strong>解释：</strong>速度为 1 时：
- 第 1 趟列车运行需要 1/1 = 1 小时。
- 由于是在整数时间到达，可以立即换乘在第 1 小时发车的列车。第 2 趟列车运行需要 3/1 = 3 小时。
- 由于是在整数时间到达，可以立即换乘在第 4 小时发车的列车。第 3 趟列车运行需要 2/1 = 2 小时。
- 你将会恰好在第 6 小时到达。
</pre>
6-（3-1）=4
2/4=0.5
max_dist


<p><strong>示例 2：</strong></p>

<pre>
<strong>输入：</strong>dist = [1,3,2], hour = 2.7
<strong>输出：</strong>3
<strong>解释：</strong>速度为 3 时：
- 第 1 趟列车运行需要 1/3 = 0.33333 小时。
- 由于不是在整数时间到达，故需要等待至第 1 小时才能搭乘列车。第 2 趟列车运行需要 3/3 = 1 小时。
- 由于是在整数时间到达，可以立即换乘在第 2 小时发车的列车。第 3 趟列车运行需要 2/3 = 0.66667 小时。
- 你将会在第 2.66667 小时到达。</pre>
2.7-(n-1)=0.7
0.7>0
2/0.7
<p><strong>示例 3：</strong></p>

<pre>
<strong>输入：</strong>dist = [1,3,2], hour = 1.9
<strong>输出：</strong>-1
<strong>解释：</strong>不可能准时到达，因为第 3 趟列车最早是在第 2 小时发车。</pre>

<p> </p>

<p><strong>提示：</strong></p>

<ul>
	<li><code>n == dist.length</code></li>
	<li><code>1 <= n <= 10<sup>5</sup></code></li>
	<li><code>1 <= dist[i] <= 10<sup>5</sup></code></li>
	<li><code>1 <= hour <= 10<sup>9</sup></code></li>
	<li><code>hours</code> 中，小数点后最多存在两位数字</li>
</ul>



## 分析
首先构造例子
dist[1,3,2]
hour=2.1
ans=20
2.1-（n-1）=0.1
2/0.1=20

In [9]:
print(2.1-2)


0.10000000000000009


### python 中四舍五入的问题
python中会有一些奇怪的问题
比如2.1-2=0.10000000000000009
最方便的四舍五入 int(X+0.5)
其他的工具round
http://www.zglg.work/data-analysis-passages/Pandas%20%E4%B8%AD%20nan%2C%20extract%2C%20round%20%E4%BD%BF%E7%94%A8%E6%80%BB%E7%BB%93/
### python中向上

In [18]:
from typing import List
import math
class Solution:
    def minSpeedOnTime(self, dist: List[int], hour: float) -> int:
        def getMaxSpeed()->int:#自带dist可以使用共有变量
            MaxSpeedPerHour=max(dist)
            MaxSpeedLastTrain=0
            lastTime=hour-(len(dist)-1)
            print(hour,hour-float(2),lastTime)
            if lastTime<=0:
                return -1
            else:
                MaxSpeedLastTrain=int(dist[-1]/lastTime+0.5)#注意这里有除磷错
                return max(MaxSpeedLastTrain,MaxSpeedPerHour)
        # print(getMaxSpeed())
        def check(speed:int):
            CostTime=0
            for i_dist in dist[:-1]:
                CostTime+=math.ceil(i_dist/speed)
            CostTime+=dist[-1]/speed
            return CostTime
        left,right=1,getMaxSpeed()
        if right==-1:
            return -1
        else:
            while left<=right:
                mid=left+((right-left)>>1)
                CostTime=check(mid)
                if CostTime>hour:#提速
                    left=mid+1
                elif CostTime<hour:
                    right=mid-1#降速
                elif CostTime==hour:
                    right=mid-1
            return left
            

# if __name__=="__main__":
if __name__=="__main__":
    solution=Solution()
    print(solution.minSpeedOnTime([1,3,2],2.1))
    print(solution.minSpeedOnTime([1,3,2],2.7))
    print(solution.minSpeedOnTime([1,3,2],6))
    print(solution.minSpeedOnTime([1,3,2],1.9))
    
    

2.1 0.10000000000000009 0.10000000000000009
20
2.7 0.7000000000000002 0.7000000000000002
3
6 4.0 4
1
1.9 -0.10000000000000009 -0.10000000000000009
-1


In [ ]:
#另一种解法
class Solution:
    def minSpeedOnTime(self, dist: List[int], hour: float) -> int:
        if hour <= len(dist)-1:
            return -1
        l=1
        '''Here r is set to 10**7 instead of max(r) only to satisfy the edge testcases'''
        r=10**7
        
        while(l<r):
            mid= l+ (r-l)//2
            sums=0
            for i in dist:
                '''Apply math.ceil to the previous value of sum because the question tells us that if a train ride takes 1.5 hours we have to wait an additional 0.5 hours'''
                sums=math.ceil(sums)+ (i)/mid
            if sums<=hour:
                r=mid
            else:
                l=mid+1
        return l